driver-core/src/main/com/mongodb/internal/TimeoutContext.java

```diff
@@ -201,11 +210,18 @@ public long getMaxAwaitTimeMS() {
 
     public long getMaxTimeMS() {
```

Comment:
This is tricky refactor, but I think we end up with?:
```
Timeout t = timeout != null 
        ? timeout 
        : Timeout.expiresIn(timeoutSettings.getMaxTimeMS(), MILLISECONDS);
t = t.shortenBy(minRoundTripTimeMS, MILLISECONDS);
return t.run(MILLISECONDS,
        () -> 0L,
        (ms) -> ms,
        () -> {
            throw createMongoRoundTripTimeoutException();
        });
```
That is, we start with the legacy `maxTimeMS` if we don't have a timeout, but whichever one we use is shortened by the round trip time (I think your snippet above does not shorten legacy time). Please confirm.

I refactored to push the run into a helper "put" method. This seemed to be the only use of maxTimeMS, which left `getMaxTimeMS` with only test usages. I can clean that up if you'd like, otherwise can be part of previously used JAVA-5385.

Reply:
If there is no timeout then you defer to `setting.getMaxTimeMS`. That figure is used as is and doesn't take round trip into account.

Reply:
Updated